Add `tsh config` helper to generate OpenSSH client configuration

timothyb89 opened the pull request:
This adds a new subcommand, `tsh config`, to generate OpenSSH client configuration snippets that allow users to connect directly to nodes using the standard `ssh` client.

To support this change, tsh's `known_hosts` file has been modified to match the format required by OpenSSH when verifying hosts against certificates. Old-style `known_hosts` entries will be automatically replaced and pruned when the end user first logs in with an updated `tsh`. Small changes were additionally made to the keystore and key agent to pass the proxy host into `AddKnownHostKeys` and to support wildcard hostnames in `known_hosts` entries.

As an example, on my machine, the generated configuration from `tsh config` looks like this:
```
#
# Begin generated Teleport configuration for proxy.foo.example.com:3080 from `tsh config`
#

# Common flags for all foo.example.com hosts
Host *.foo.example.com proxy.foo.example.com
    UserKnownHostsFile "/Users/tim/.tsh/known_hosts"

# Flags for all foo.example.com hosts except the proxy
Host *.foo.example.com !proxy.foo.example.com
    Port 3022
    ProxyCommand ssh -p 3023 proxy.foo.example.com -s proxy:$(echo %h | cut -d '.' -f 1):%p

# Common flags for all leaf.example.com hosts
Host *.leaf.example.com proxy.foo.example.com
    UserKnownHostsFile "/Users/tim/.tsh/known_hosts"

# Flags for all leaf.example.com hosts except the proxy
Host *.leaf.example.com !proxy.foo.example.com
    Port 3022
    ProxyCommand ssh -p 3023 proxy.foo.example.com -s proxy:$(echo %h | cut -d '.' -f 1):%p@leaf.example.com

# End generated Teleport configuration
```

... and the revised `~/.tsh/known_hosts` format looks like this:
```
@cert-authority proxy.foo.example.com,foo.example.com,*.foo.example.com ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCt6aILUzSic4xq/BmrLsIZNs2wfzNbVE0Qlft2dUo4Sb9mwODVhvYLDMfgI1nWsDCTcwWHbu1o0j6GxLhnyigEHJpavpBbn4YqutTQ1xp35dWMi76PSUu7bNCAx3IZz8E1JhELHzFUSFAY5CuyuIy+cxKsD+CdN0xFUNtjqozX6BCp8bcY+WQCNLDf8H1m8RST/FeTaBmPtBzxF6yei7ehczEcBTC1czbprQ7piiBI6CGNeE9zKdKYHDtbQyRGs8Cy56jCF0fgEdXO0p7DqJIcr0/IjmdW6OQku1MKC8v3VxusXCVxKuJ9T5TeiRPglyW6APy4iKZYAO40Wy2WONcf type=host
@cert-authority proxy.foo.example.com,leaf.example.com,*.leaf.example.com ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDWUE468X1slCu5xGDIZInkVdLRIi1FU6x6221+ecr9h1KVBP0flzSssRBS7Qfa47uJ+cJaak3/luwnUmIgtQX4ovU4r5MJuZhaAreYMF6A73vDMehhOUH3KsNkEvMwOlOREJYt2YM8RZ6eSLgyo/HDlX8BHfl4uKDd9qOqvKSNab5VDstEHuoslEEj8YG2v8FaBMKUjudheSzmzRvdeSKVy9FjHA5yZslXzcMS0S0QACL4BleWwzpp/zZ++xGiTbtdb50FKhrc7REQQHmec/O/N31mLyzsUnJjeBcipphX/iYVnnzgvY78N7RfBnebNmt4To1KCul1HVK9jgsqcrUf type=host
```

Fixes https://github.com/gravitational/teleport/issues/3734

timothyb89 committed:
Add `tsh config ssh` helper to generate OpenSSH client configuration

This adds a new subcommand, `tsh config ssh`, to generate OpenSSH
client configuration snippets that allow users to connect directly to
nodes using the standard `ssh` client.

To support this change, tsh's `known_hosts` file has been modified to
match the format required by OpenSSH when verifying hosts against
certificates. Old-style `known_hosts` entries will be automatically
replaced and pruned when the end user first logs in with an updated
`tsh`. Small changes were additionally made to the keystore and key
agent to pass the proxy host into `AddKnownHostKeys` and to support
wildcard hostnames in `known_hosts` entries.

timothyb89 commented:
Hi all, I've got a few open questions I'd especially like some feedback on here:

 1. This mildly changes the semantics of tsh's keystore. Previously, tsh only wrote a single hostname, `clusterName`, into each `known_hosts` entry. With this change, entries need to have at least `*.clusterName` and `proxyHost` in their hosts list, and with just those two both OpenSSH and `tsh ssh` appear to work properly. However:
     * It breaks the semantics of the `GetKnownHostKeys` - if I insert `example.com`, I could only retrieve it with `*.example.com` (or the proxy host, though occasionally the `proxyHost` and `clusterName` are identical)
     * Because of this, 2 unit tests break. They could be fixed by retrieving `*.example.com` instead.
     * I'm unclear if tsh ever depends on bare `clusterName` retrieval.

    Given this, I've opted to additionally include the plain `clusterName` in each `known_hosts` entry. Is this the right thing to do?
 2. As we're now inserting wildcard hosts, `GetKnownHostKeys` could only retrieve a literal `*.example.com` and would not match, say, `foo.example.com`. I have modified `fsLocalNonSessionKeyStore.GetKnownHostKeys()` to additionally match simple wildcard hostnames, however I don't believe this is used anywhere. While it does resolve a soundness issue, it does increase attack surface. Is this a change we want to keep?
 3. Should we stick with `tsh config ssh` or instead use @r0mant's suggestion in https://github.com/gravitational/teleport/issues/3734#issuecomment-866560974 to use just `tsh config`?

timothyb89 committed:
Fix broken link to Trusted Clusters documentation

russjones commented on lib/client/known_hosts_migrate.go:
```diff
@@ -0,0 +1,159 @@
+	// text to preserve formatting for all passed-through entries.
+	marker, hosts, pubKey, comment, _, err := ssh.ParseKnownHosts([]byte(raw))
+	if err != nil {
+		return nil, err
```
Use `trace.Wrap` when returning errors, in debug mode this will give us a stack trace of where the error occured.

timothyb89 replied:
Sure, fixed.

russjones commented on tool/tsh/config.go:
```diff
@@ -0,0 +1,105 @@
+
+func writeSSHConfig(sb *strings.Builder, clusterName string, knownHostsPath string, proxyHost string, proxyPort string, leaf bool) {
+	// Generate configuration for all Teleport targets.
+	fmt.Fprintf(sb, "# Common flags for all %s hosts\n", clusterName)
```
Any reason you didn't use https://golang.org/pkg/text/template/ here?

xacrimon replied:
Just a note on this from me. IMO the template package lends itself more to things that are actual templates, i.e they get used multiple times per program run with changing parameters. For one-off string interpolation it's usually slower and less ergonomic.

timothyb89 replied:
No real reason on my end, so I've just rewritten the function to use a template. It's presumably slower but will only ever be invoked a handful of times so I don't expect it to be very costly. I'm happy to revert if there's strong feelings one way or the other.

russjones commented on tool/tsh/tsh.go:
```diff
@@ -499,6 +499,17 @@ func Run(args []string, opts ...cliOption) error {
 	// MFA subcommands.
 	mfa := newMFACommand(app)
 
+	config := app.Command("config", "Manage application configuration")
```
@klizhentas I don't think we reached a consensus in the ticket, this was the last comment from @r0mant: https://github.com/gravitational/teleport/issues/3734#issuecomment-866560974

timothyb89 replied:
I've just pushed a change to to make it just `tsh config`.

timothyb89 committed:
Use text/template for SSH config generation; wrap all errors.

webvictim reviewed (commented):
Just a couple of general implementation questions:
1) How does this work with IP addresses? A lot of customers/PoCs still use `ssh root@10.1.2.3` - will this handle IP ranges OK?
2) What happens if `proxy_service.public_addr` isn't set for a given cluster?

timothyb89 commented:
> 1. How does this work with IP addresses? A lot of customers/PoCs still use `ssh root@10.1.2.3` - will this handle IP ranges OK?

They won't work as-is, since both the config generation and known_hosts depend on having some predictable way of addressing nodes for a particular cluster. The existing [OpenSSH client guide](https://goteleport.com/docs/server-access/guides/openssh/#use-openssh-client) notes a similar limitation however as far as I can tell the node host certificates have their IPs listed as principals as well.

It could probably be made to work if we listed out every individual node IP in both the generated config and in `~/.tsh/known_hosts`, but I'm not sure how I feel about doing that at this stage.

Maybe this is a good candidate to implement in a future pass, alongside `Match exec` host matching? I'll also note that recent versions of OpenSSH support a [`KnownHostsCommand`](https://man.openbsd.org/ssh_config#KnownHostsCommand) directive which could generate a proper known hosts entry on the fly if needed.

> 2. What happens if `proxy_service.public_addr` isn't set for a given cluster?

It still generates the same output on my test cluster, I assume the client falls back to the address as written by the user if the public address is unset?

timothyb89 committed:
Rename config helper from `config ssh` to just `config`

This changes the config helper to use just `tsh config` per
suggestion from @r0mant.

timothyb89 committed:
Fix known_hosts_migrate_test after rebase

timothyb89 committed:
Merge branch 'master' into timothyb89/ssh-config-helper

russjones commented:
@r0mant  @xacrimon @jane Can you review?

r0mant commented on lib/client/keystore.go:
```diff
@@ -557,6 +572,28 @@ func (fs *fsLocalNonSessionKeyStore) AddKnownHostKeys(hostname string, hostKeys
+		return false
+	}
+
+	hostnameParts := strings.Split(hostname, ".")
+	hostnameRoot := strings.Join(hostnameParts[1:], ".")
```
How will this behave if `hostname` has 1 part?

timothyb89 replied:
It behaves sensibly (will effectively never match unless the pattern is the empty string) however I've now made this an explicit requirement (hostnames must contain `.`) and added a testcase.

r0mant commented on lib/client/keystore.go:
```diff
@@ -581,6 +618,9 @@ func (fs *fsLocalNonSessionKeyStore) GetKnownHostKeys(hostname string) ([]ssh.Pu
 					if hosts[i] == hostname {
 						hostMatch = true
 						break
+					} else if matchesWildcard(hostname, hosts[i]) {
+						hostMatch = true
+						break
```
```suggestion
					if hosts[i] == hostname || matchesWildcard(hostname, hosts[i]) {
						hostMatch = true
						break
```

r0mant commented on lib/client/keystore.go:
```diff
@@ -666,7 +706,7 @@ func (noLocalKeyStore) DeleteUserCerts(idx KeyIndex, opts ...CertOption) error {
 }
 func (noLocalKeyStore) DeleteKeys() error { return errNoLocalKeyStore }
-func (noLocalKeyStore) AddKnownHostKeys(hostname string, keys []ssh.PublicKey) error {
+func (noLocalKeyStore) AddKnownHostKeys(hostname string, proxyHost string, keys []ssh.PublicKey) error {
```
```suggestion
func (noLocalKeyStore) AddKnownHostKeys(hostname, proxyHost string, keys []ssh.PublicKey) error {
```

r0mant commented on lib/client/known_hosts_migrate.go:
```diff
@@ -0,0 +1,159 @@
+	// text to preserve formatting for all passed-through entries.
+	marker, hosts, pubKey, comment, _, err := ssh.ParseKnownHosts([]byte(raw))
+	if err != nil {
+		return nil, err
```
```suggestion
		return nil, trace.Wrap(err)
```

r0mant commented on lib/client/known_hosts_migrate.go:
```diff
@@ -0,0 +1,159 @@
+	}
+
+	return &knownHostEntry{
+		raw,
+		marker,
+		hosts,
+		pubKey,
+		comment,
```
Can you name the fields explicitly please?

timothyb89 replied:
Sure, done.

r0mant commented on lib/client/known_hosts_migrate.go:
```diff
@@ -0,0 +1,159 @@
+
+	// Note: we already know this entry matches exactly 1 host. The new-style
+	// entries prepend `*.`, so we'll add that upfront.
+	oldHost := "*." + oldEntry.hosts[0]
```
Do you need to check `hosts` length? To account for malformed entries and such.

timothyb89 replied:
This is checked in `isOldStyleHostsEntry()`; `canPruneOldHostsEntry()` is only called on entries for which `isOldStyleHostsEntry` returned true. 

In any case, I've added an additional `isOldStyleHostsEntry()` check here so the function will at least never panic if given hostile arguments.

r0mant commented on lib/client/known_hosts_migrate.go:
```diff
@@ -0,0 +1,159 @@
+	// exists. If not, pass it through.
+	for _, entry := range oldEntries {
+		if canPruneOldHostsEntry(entry, newEntries) {
+			log.Debugf("Pruning old known_hosts entry for %s.", entry.hosts[0])
```
Check hosts length?

timothyb89 replied:
Old entries have their length checked in `isOldStyleHostsEntry()` and are guaranteed to have exactly 1 host entry. Additionally,  `crypto/ssh`'s `ParseKnownHosts` rejects lines without any hosts so the list should never be empty to begin with.

If this is still problematic for code style reasons I'm happy to add more checks, though.

r0mant commented on lib/client/known_hosts_migrate.go:
```diff
@@ -0,0 +1,159 @@
+		parsed, err := parseKnownHost(line)
+		if err != nil {
+			// If the line isn't parseable, pass it through.
+			log.Debugf("Unable to parse known host on line %d, skipping", i+1)
```
```suggestion
			log.WithError(err).Debugf("Unable to parse known host on line %d, skipping.", i+1)
```

r0mant commented on tool/tsh/config.go:
```diff
@@ -0,0 +1,126 @@
+Host *.{{ .clusterName }} !{{ .proxyHost }}
+    Port 3022
+    {{- if .leaf }}
+    ProxyCommand ssh -p {{ .proxyPort }} {{ .proxyHost }} -s proxy:$(echo %h | cut -d '.' -f 1):%p@{{ .clusterName }}	
```
Can `$(echo %h | cut -d '.' -f 1)` be done by Go code instead and pasted here as a template variable?

Same below.

timothyb89 replied:
It is possible but would require another subcommand of some sort to be built into tsh. Per advice from @russjones we've decided to defer for a bit as the upcoming `tsh connect` functionality could supersede any text manipulation we do.

r0mant commented on tool/tsh/config.go:
```diff
@@ -0,0 +1,126 @@
+
+// writeSSHConfig generates an OpenSSH config block from the `sshConfigTemplate`
+// template string.
+func writeSSHConfig(
+	sb *strings.Builder, clusterName string, knownHostsPath string,
+	proxyHost string, proxyPort string, leaf bool,
+) error {
```
```suggestion
func writeSSHConfig(sb *strings.Builder, clusterName, knownHostsPath, proxyHost, proxyPort string, leaf bool) error {
```

r0mant commented on tool/tsh/config.go:
```diff
@@ -0,0 +1,126 @@
+		"leaf":           leaf,
+	})
+	if err != nil {
+		return trace.WrapWithMessage(err, "Error generating SSH configuration from template")
```
```suggestion
		return trace.Wrap(err, "error generating SSH configuration from template")
```

timothyb89 committed:
First pass at review feedback

timothyb89 committed:
Update docs/pages/server-access/guides/openssh.mdx

Co-authored-by: Roman Tkachenko <roman@gravitational.com>

timothyb89 committed:
Ensure top-level hostnames never match wildcard patterns

timothyb89 committed:
Merge branch 'timothyb89/ssh-config-helper' of github.com:gravitational/teleport into timothyb89/ssh-config-helper

timothyb89 committed:
Add additional host count check to `canPruneOldHostsEntry`.

timothyb89 committed:
Merge branch 'master' into timothyb89/ssh-config-helper

xacrimon commented on lib/client/known_hosts_migrate.go:
```diff
@@ -0,0 +1,163 @@
+	// only safe to prune an old entry when both the (*.)hostname and public key
+	// match.
+
+	// Ensure that (among other conditions) this entry matches exactly 1 host.
+	if !isOldStyleHostsEntry(oldEntry) {
+		return false
+	}
```
You can usually avoid this check since it's already done in the caller but only if you explicitly note the invariants in the documentation of `canPruneOldHostsEntry`

timothyb89 replied:
Fine by me, done

xacrimon reviewed (commented):
LGTM otherwise on the implementation

timothyb89 committed:
Replace excess call to `isOldStyleHostsEntry` with documented invariant

timothyb89 committed:
Merge branch 'master' into timothyb89/ssh-config-helper

timothyb89 committed:
Trim trailing dots on absolute hostnames in `matchesWildcard`

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/ssh-config-helper

timothyb89 commented on lib/client/keystore.go:
```diff
@@ -557,6 +572,33 @@ func (fs *fsLocalNonSessionKeyStore) AddKnownHostKeys(hostname string, hostKeys
+// The `pattern` may be prefixed with `*.` which will match exactly one domain
+// segment, meaning `*.example.com` will match `foo.example.com` but not
+// `foo.bar.example.com`.
+func matchesWildcard(hostname, pattern string) bool {
```
Good catch, fixed

webvictim replied:
https://github.com/gravitational/teleport/issues/5439 actually tracks a very similar issue, so at least one person is trying it :)

russjones reviewed (approved):
Bot.

timothyb89 committed:
Merge branch 'master' into timothyb89/ssh-config-helper
